torch/csrc/utils/tensor_new.cpp

```diff
@@ -115,8 +115,9 @@ Tensor new_with_storage(c10::TensorTypeId type_id, at::ScalarType scalar_type, S
 }
 
 Tensor new_with_tensor(c10::TensorTypeId type_id, at::ScalarType scalar_type, const Tensor& other) {
-  if (other.type_id() != type_id) {
-    throw TypeError("expected %s (got %s)", type_id, other.type_id());
+  if (!other.type_set().has(type_id)) {
```

Comment:
`legacyExtract` would have been suitable here. In fact, since this is legacy code, I should probably use `legacyExtract` here.

The reasoning for why TensorTypeId equality test should turn into set equality is something like this:

1. We want to be semantics preserving
2. Eventually, some of these type sets are going to contain variable. But the old test would have still passed even if the tensor was a variable. So id equality to set equality is wrong
3. What's a set operation that does work in this case? Set membership.

When I wrote the patch, I didn't have `legacyExtract`; eventually I added it because some of the cases were too annoying to reason out.

Reply:
The semantic difference between these two, is that `set.has(VariableTensorId)` can succeed, but `legacyExtract(set) == VariableTensorId` is always false.